dra-evolution/pkg/api/claim_types.go

```diff
@@ -1,379 +1,679 @@
+	// on a node which match the requrirements.
 	//
-	// NOTE: This cannot appear in class because the quota code does not do
-	// an indirection. Still searching for a better way to handle this use
-	// case, without creating a new top-level type for it.
+	// Default is false.
 	//
 	// +optional
 	AdminAccess *bool `json:"adminAccess,omitempty"`
 
-	// Constraints is a CEL expression that operates on device attributes.
-	// In order for a device to be considered, this CEL expression and the
-	// Constraints expression from the DeviceClass must both be true.
+	// All these match criteria must be satisfied by all devices allocated for this request.
+	//
+	// +optional
+	// +listType=atomic
+	Match []MatchModel `json:"match,omitempty"`
+
+	// Count defines how many instances are desired. If unset, exactly one
+	// instance must be available. When a range is set, it is possible to
+	// ask for:
+	// - x >= minimum instances up to all that are available
+	// - 0 <= x <= maximum (up to a certain number, with zero instances acceptable)
+	// - minimum <= 0 <= maximum (within a certain range)
+	// +optional
+	Count *CountDetails `json:"count,omitempty"`
+
+	// Requirements describe additional contraints that all must be met by a device
+	// to satisfy the request.
 	//
 	// +optional
-	Constraints *string `json:"constraints,omitempty"`
+	// +listType=atomic
+	Requirements []Requirement `json:"requirements,omitempty" protobuf:"bytes,4,opt,name=requirements"`
+}
+
+// CountDetails defines how many instances are desired.
+type CountDetails struct {
+	// Minimum defines the lower limit. At least this many instances
+	// must be available (x >= minimum). The default if unset is one.
+	Minimum *int `json:"minimum"`
 
-	// Device claims may be satisfied by choosing multiple devices instead
-	// of just a single device.
+	// Maximum defines the upper limit. At most this many instances
+	// may be allocated (x <= maximum). The default if unset is unlimited.
+	Maximum *int `json:"maximum"`
+}
 
-	// MaxDevices allows the user to specify a maximum number of devices
-	// that may be allocated to sastisfy this claim. The default is equal
-	// to requests.devices. Thus, without specifying MaxDevices, the user
-	// will get exactly the number of devices specified in
-	// requests.devices.
+// MatchModel must have one and only one field set.
+type MatchModel struct {
+	// All devices must have this attribute and its value must be the same.
+	//
+	// For example, if you specified "numa.dra.example.com" (a hypothetical example!),
+	// then only devices in the same NUMA node will be chosen.
 	//
 	// +optional
-	MaxDevices *int `json:"maxDevices,omitempty"`
+	// +listType=atomic
+	Attribute *string `json:"attribute,omitempty"`
 
-	// Requests allows the user to specify the minimum resource
-	// requirements that must be satisfied across all devices allocated for
-	// this claim.
+	// Future extension, not part of the current design:
```

Comment:
I was thinking of giving the CEL expression access to all devices and then letting it evaluate whether those match. I know that there are list (reduce?) operations, but whether those are sufficient is completely open. Perhaps we have to do pair-wise comparisons, but that would be less expressive.

Reply:
It would be useful to have something like what @thockin describes (at least until we standardize on some of the attributes). That way I can say, e.g. `device.stringAttributes['pcie-root.gpu.nvidia.com'] == device.stringAttributes['pcie-root.rdma.nvidia.com']`. Otherwise we have no way to do this across drivers.

Reply:
Useful, yes. Required, perhaps not.